repo2docker/contentproviders/dataverse.py

```diff
@@ -0,0 +1,129 @@
+            # Knowing the file identifier query search api to get parent dataset
+            search_url = urlunparse(
+                parsed_url._replace(path="/api/search", query=search_query)
+            )
+            resp = self.urlopen(search_url).read()
+            data = json.loads(resp.decode("utf-8"))["data"]
+            if data["count_in_response"] != 1:
+                self.log.debug("Dataverse search query failed!")
+                self.log.debug("  - doi = " + doi)
+                self.log.debug("  - url = " + url)
+                self.log.debug("  - resp = " + json.dumps(data))
+                return
+
+            self.record_id = deep_get(data, "items.0.dataset_persistent_id")
+        elif is_doi(doi):
+            self.record_id = "doi:" + normalize_doi(doi)
+
+        if hasattr(self, "record_id"):
+            return {"record": self.record_id, "host": host}
+
+    def fetch(self, spec, output_dir, yield_output=False):
+        """Fetch and unpack a Dataverse dataset."""
+        record_id = spec["record"]
+        host = spec["host"]
+
+        yield "Fetching Dataverse record {}.\n".format(record_id)
+        req = Request(
+            "{}/api/datasets/:persistentId?persistentId={}".format(
+                host["url"], record_id
+            ),
+            headers={"accept": "application/json"},
+        )
+        resp = self.urlopen(req)
+        record = json.loads(resp.read().decode("utf-8"))["data"]
+
+        # In order to fetch entire dataset we build a list of file IDs we want to fetch
+        # and then receive a zip file containing all of them.
+        # TODO: Dataverse has a limit for the zipfile size (see
+        # https://github.com/jupyter/repo2docker/pull/739#issuecomment-510834729)
+        # If size of the dataset is grater than 100MB individual files should be downloaded.
+        file_ids = [
+            str(deep_get(fobj, "dataFile.id"))
+            for fobj in deep_get(record, "latestVersion.files")
+        ]
+
+        req = Request(
+            "{}/api/access/datafiles/{}".format(host["url"], ",".join(file_ids))
+        )
+
+        dst_fname = os.path.join(output_dir, "dataverse.zip")
+        with self.urlopen(req) as src, open(dst_fname, "wb") as dst:
+            yield "Fetching files bundle\n"
+            shutil.copyfileobj(src, dst)
+
+        yield "Extracting files\n"
+        with ZipFile(dst_fname) as zfile:
+            zfile.extractall(path=output_dir)
+
+        os.remove(dst_fname)
+        new_subdirs = os.listdir(output_dir)
+        # if there is only one new subdirectory move its contents
+        # to the top level directory
+        if len(new_subdirs) == 1:
+            d = new_subdirs[0]
+            copytree(os.path.join(output_dir, d), output_dir)
+            shutil.rmtree(os.path.join(output_dir, d))
+
+    @property
+    def content_id(self):
+        """The Dataverse persistent identifier (could use internal dataset_id too)."""
```

Comment:
If you stick to the `doi:` prefix, suggest to mention that here. Also, remove the "could use internal dataset_id", if that is not used eventually - see comment above.